glue/viewers/image/composite_array.py

```diff
@@ -80,7 +82,44 @@ def __call__(self, bounds=None):
         img = None
         visible_layers = 0
 
-        for uuid in sorted(self.layers, key=lambda x: self.layers[x]['zorder']):
+        # We first check that layers are either all colormaps or all single
+        # colors, and at the same time, if we use colormaps we ccheck what
+        # the smallest alpha value is - if it is 1 then colormaps do not change
+        # transparency and this allows us to speed things up a little.
+
+        minimum_cmap_alpha = 1
+        mode = None
+
+        for layer in self.layers.values():
+            if isinstance(layer['color'], Colormap):
+                if mode is None:
+                    mode = 'colormap'
+                elif mode == 'color':
+                    raise TypeError("Mixed color and colormaps in CompositeArray")
```

Comment:
How often would someone encounter this now that it throws exception? I don't think there is a mechanism to prevent user from setting the first viewer layer to colormap but the second layer to monochromatic?

Also, are things like subsets and linked tables already excluded when this logic is invoked? I am not familiar with the internals here.

Reply:
Glue doens't actually allow image viewers to have mixed layers - the mode is controlled by:

https://github.com/glue-viz/glue/blob/e4e875a7a54c4925db244f62f7f16644578617dd/glue/viewers/image/state.py#L89-L92

In fact I should make CompositeArray have such a settable mode rather than checking it like this at drawing time - but that can be cleaned up/improved in future. For now, these exceptions should never be tripped by users and if they are it indicates a bug in the image viewers.

Reply:
The speed up is looking very promising!  But I am seeing this error raised when adding data to the viewer if the viewer itself is visible.  Let me know if you need the full traceback or more details to reproduce.

Reply:
Could you provide the full traceback?

Reply:
```
File ~/repos/jdaviz/jdaviz/app.py:1380, in Application.set_data_visibility(self, viewer_reference, data_label, visible, replace)
   1378 [data] = [x for x in self.data_collection if x.label == data_label]
   1379 color_cycler = ColorCycler()
-> 1380 viewer.add_data(data, percentile=95)
   1382 add_data_message = AddDataMessage(data, viewer,
   1383                                   viewer_id=viewer_id,
   1384                                   sender=self)
   1385 self.hub.broadcast(add_data_message)

File /opt/miniconda3/envs/jdaviz-dev/lib/python3.8/site-packages/glue_jupyter/view.py:117, in IPyWidgetView.add_data(self, data, color, alpha, **layer_state)
    113 def add_data(self, data, color=None, alpha=None, **layer_state):
    115     data = validate_data_argument(self._data, data)
--> 117     result = super().add_data(data)
    119     if not result:
    120         return

File ~/repos/glue/glue/viewers/common/viewer.py:218, in Viewer.add_data(self, data)
    216 with ignore_callback(layer.state, 'zorder'):
    217     self._layer_artist_container.append(layer)
--> 218 layer.update()
    219 self.draw_legend()  # need to be called here because callbacks are ignored in previous step
    221 # Add existing subsets to viewer

File ~/repos/glue/glue/utils/matplotlib.py:181, in defer_draw.<locals>.wrapper(*args, **kwargs)
    177 @wraps(func)
    178 def wrapper(*args, **kwargs):
    180     if len(DEFER_DRAW_BACKENDS) == 0:
--> 181         return func(*args, **kwargs)
    183     # Don't recursively defer draws. We just check the first draw_idle
    184     # method since all should be modified in sync.
    185     if isinstance(DEFER_DRAW_BACKENDS[0].draw_idle, DeferredMethod):

File ~/repos/glue/glue/viewers/image/layer_artist.py:202, in ImageLayerArtist.update(self, *event)
    200 ARRAY_CACHE.pop(self.state.uuid, None)
    201 PIXEL_CACHE.pop(self.state.uuid, None)
--> 202 self._update_image(force=True)
    203 self.redraw()

File /opt/miniconda3/envs/jdaviz-dev/lib/python3.8/site-packages/glue_jupyter/bqplot/image/layer_artist.py:127, in BqplotImageLayerArtist._update_image(self, force, **kwargs)
    123 changed = self.pop_changed_properties()
    125 if force or any(prop in changed for prop in ('layer', 'attribute',
    126                                              'slices', 'x_att', 'y_att')):
--> 127     self._update_image_data()
    128     force = True  # make sure scaling and visual attributes are updated
    130 if force or any(prop in changed for prop in ('v_min', 'v_max', 'contrast',
    131                                              'bias', 'alpha', 'color_mode',
    132                                              'cmap', 'color', 'zorder',
    133                                              'visible', 'stretch',
    134                                              'bitmap_visible', 'contour_visible')):

File /opt/miniconda3/envs/jdaviz-dev/lib/python3.8/site-packages/glue_jupyter/bqplot/image/layer_artist.py:140, in BqplotImageLayerArtist._update_image_data(self, *args, **kwargs)
    139 def _update_image_data(self, *args, **kwargs):
--> 140     self.composite_image.invalidate_cache()
    141     # if the image data change, the contour lines are invalid
    142     self._contour_line_cache.clear()

File /opt/miniconda3/envs/jdaviz-dev/lib/python3.8/site-packages/glue_jupyter/bqplot/image/frb_mark.py:107, in FRBImage.invalidate_cache(self)
    106 def invalidate_cache(self):
--> 107     self.update()

File /opt/miniconda3/envs/jdaviz-dev/lib/python3.8/site-packages/glue_jupyter/bqplot/image/frb_mark.py:95, in FRBImage.update(self, *args, **kwargs)
     92 bounds = [(ymin, ymax, ny), (xmin, xmax, nx)]
     94 # Get the array and assign it to the artist
---> 95 image = self.array_maker(bounds=bounds)
     96 if image is not None:
     97     if image.dtype == np.dtype("float64"):

File ~/repos/glue/glue/viewers/image/composite_array.py:106, in CompositeArray.__call__(self, bounds)
    104             mode = 'color'
    105         elif mode == 'colormap':
--> 106             raise TypeError("Mixed color and colormaps in CompositeArray")
    108 # Get a sorted list of UUIDs with the top layers last
    109 sorted_uuids = sorted(self.layers, key=lambda x: self.layers[x]['zorder'])

TypeError: Mixed color and colormaps in CompositeArray
```

Reply:
Aww, man.

`viewer.add_data(data, percentile=95)` -- I added the `percentile` in https://github.com/spacetelescope/jdaviz/pull/1386 but the `add_data` call has always been there.

Reply:
we now pass a default `color` as well, but I temporarily removed it locally in case that was triggering the error, but that doesn't seem to be the case.